frontend/src/toolbar/elements/heatmapLogic.ts

```diff
@@ -137,12 +147,10 @@ export const heatmapLogic = kea<heatmapLogicType>([
         ],
     })),
 
-    selectors({
+    selectors(({ cache }) => ({
         elements: [
-            (selectors) => [selectors.elementStats, toolbarLogic.selectors.dataAttributes],
-            (elementStats, dataAttributes) => {
-                // cache all elements in shadow roots
-                const allElements = collectAllElementsDeep('*', document)
+            (selectors) => [selectors.elementStats, toolbarLogic.selectors.dataAttributes, selectors.pageElements],
+            (elementStats, dataAttributes, pageElements) => {
```

Comment:
Could this just depend on the `href` and avoid all this logic around getting and setting `pageElements`? If you're anyway using a `cache`, you could just bust it here if the URL changed.

Reply:
Not understanding, sorry... This elements selector has to depend on those things since that's how we get from the API result to the elements. Maybe easier to clarify sync?

(I also wanted to change as little as possible since there are no tests on this 🙈 )

Reply:
Putting the href in the selector array assures that this selector gets recalculated every time. I mean something like this instead of introducing the new actions and reducers

```
            (selectors) => [selectors.elementStats, toolbarLogic.selectors.dataAttributes, currentPageLogic.selectors.href],
            (elementStats, dataAttributes, href) => {
                const pageElements = cache.lastHref == href ? {} : cache.pageElements
                const selectorToElements = cache.lastHref == href ? {} : cache.selectorToElements

                cache.pageElements = pageElements
                cache.selectorToElements = selectorToElements
                cache.lastHref = href
                // ..
            }
```